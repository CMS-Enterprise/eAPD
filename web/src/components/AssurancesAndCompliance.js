import React from 'react';

import { t } from '../i18n';
import Section from './Section';
import SectionDesc from './SectionDesc';
import SectionTitle from './SectionTitle';
import Collapsible from './Collapsible';

const AssurancesAndCompliance = () => (
  <Section id="assurances-compliance">
    <SectionTitle>{t('assurancesAndCompliance.title')}</SectionTitle>
    <SectionDesc>{t('assurancesAndCompliance.helpText')}</SectionDesc>
    <Collapsible title="Procurement Standards">
      <div>...</div>
    </Collapsible>
    <Collapsible title="Access to Records">
      <div>...</div>
    </Collapsible>
    <Collapsible title="Software Rights">
      <div>...</div>
    </Collapsible>
    <Collapsible title="Security">
      <div>...</div>
    </Collapsible>
  </Section>
);

export default AssurancesAndCompliance;
