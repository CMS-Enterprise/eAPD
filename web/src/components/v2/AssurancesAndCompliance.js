import React from 'react';

import Section from './Section';
import SectionTitle from './SectionTitle';
import Collapsible from '../Collapsible';

const AssurancesAndCompliance = () => (
  <Section>
    <SectionTitle>Assurances and Compliance</SectionTitle>
    <Collapsible title="Procurement Standards">
      <div>...</div>
    </Collapsible>
    <Collapsible title="Access to Records">
      <div>...</div>
    </Collapsible>
    <Collapsible title="Software Rights">
      <div>...</div>
    </Collapsible>
    <Collapsible title="Security">
      <div>...</div>
    </Collapsible>
  </Section>
);

export default AssurancesAndCompliance;
