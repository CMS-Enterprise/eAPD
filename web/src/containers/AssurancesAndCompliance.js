import { Choice, TextField } from '@cmsgov/design-system-core';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';

import Waypoint from './ConnectedWaypoint';
import { updateApd as updateApdAction } from '../actions/apd';
import { Section, Subsection } from '../components/Section';
import regLinks from '../data/assurancesAndCompliance.yaml';
import { t } from '../i18n';

const namify = (name, title) =>
  `explanation-${name}-${title}`.replace(/\s/g, '_');

const LinkOrText = ({ link, title }) => {
  if (!link) return title;

  return (
    <a href={link} target="_blank" rel="noopener noreferrer">
      {title}
    </a>
  );
};

LinkOrText.propTypes = {
  link: PropTypes.string,
  title: PropTypes.string.isRequired
};

LinkOrText.defaultProps = {
  link: null
};

class AssurancesAndCompliance extends Component {
  handleCheckChange = (section, index, newValue) => () => {
    const { updateApd } = this.props;
    updateApd({
      federalCitations: { [section]: { [index]: { checked: newValue } } }
    });
  };

  handleExplanationChange = (section, index) => e => {
    const { updateApd } = this.props;
    updateApd({
      federalCitations: {
        [section]: { [index]: { explanation: e.target.value } }
      }
    });
  };

  render() {
    const { sections: apdSections } = this.props;

    return (
      <Waypoint id="assurances-compliance">
        <Section
          isNumbered
          id="assurances-compliance"
          resource="assurancesAndCompliance"
        >
          <Subsection
            id="assurances-compliance-fed-citations"
            resource="assurancesAndCompliance.citations"
          >
            {Object.entries(regLinks).map(([name, regulations]) => (
              <div key={name} className="ds-u-margin-bottom--3">
                <h4 className="ds-h4">{t(`assurancesAndCompliance.headings.${name}`)}</h4>
                {apdSections[name].map(
                  ({ title, checked, explanation }, index) => (
                    <fieldset key={title} className="ds-u-margin-top--2">
                      <legend className="ds-c-label">
                        Are you complying with{' '}
                        <strong>
                          <LinkOrText link={regulations[title]} title={title} />
                        </strong>
                        ?
                      </legend>
                      <Choice
                        type="radio"
                        value="yes"
                        name={`apd-assurances-yes-${namify(name, title)}`}
                        size="small"
                        checked={checked === true}
                        onChange={this.handleCheckChange(name, index, true)}
                      >
                        Yes
                      </Choice>
                      <Choice
                        type="radio"
                        value="no"
                        name={`apd-assurances-no-${namify(name, title)}`}
                        size="small"
                        checked={checked === false}
                        onChange={this.handleCheckChange(name, index, false)}
                        checkedChildren={
                          <div className="ds-c-choice__checkedChild">
                            <TextField
                              label="Please explain"
                              name={namify(name, title)}
                              value={explanation}
                              onChange={this.handleExplanationChange(
                                name,
                                index
                              )}
                              multiline
                              rows={5}
                            />
                          </div>
                        }
                      >
                        No
                      </Choice>
                    </fieldset>
                  )
                )}
              </div>
            ))}
          </Subsection>
        </Section>
      </Waypoint>
    );
  }
}

AssurancesAndCompliance.propTypes = {
  sections: PropTypes.object.isRequired,
  updateApd: PropTypes.func.isRequired
};

const mapStateToProps = ({
  apd: {
    data: { federalCitations: sections }
  }
}) => ({ sections });

const mapDispatchToProps = {
  updateApd: updateApdAction
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AssurancesAndCompliance);

export {
  AssurancesAndCompliance as plain,
  mapStateToProps,
  mapDispatchToProps,
  LinkOrText
};
