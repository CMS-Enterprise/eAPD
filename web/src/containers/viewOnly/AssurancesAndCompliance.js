import PropTypes from 'prop-types';
import React, { Fragment } from 'react';
import { connect } from 'react-redux';

import { CheckCircle, TimesCircle } from '../../components/Icons';
import regLinks from '../../data/assurancesAndCompliance.yaml';
import { selectFederalCitations } from '../../reducers/apd.selectors';
import { t } from '../../i18n';

const AssurancesAndCompliance = ({ citations }) => {
  return (
    <div>
      <h2>Assurances and Compliance</h2>
      {Object.entries(regLinks).map(([name]) => (
        <div key={name} className="ds-u-margin-bottom--3">
          <h4 className="ds-h4">
            {t(`assurancesAndCompliance.headings.${name}`)}
          </h4>
          <ul>
            {citations[name].map(({ title, checked, explanation }) => (
              <li key={title}>
                {checked ? <CheckCircle /> : <TimesCircle />} {title}
                {!checked && (
                  <Fragment>
                    <br />
                    <strong>Explanation:</strong>{' '}
                    {explanation || <em>No explanation given</em>}
                  </Fragment>
                )}
              </li>
            ))}
          </ul>
        </div>
      ))}
    </div>
  );
};

AssurancesAndCompliance.propTypes = {
  citations: PropTypes.object.isRequired
};

const mapStateToProps = state => ({ citations: selectFederalCitations(state) });

export default connect(
  mapStateToProps,
  null
)(AssurancesAndCompliance);
export { AssurancesAndCompliance as plain, mapStateToProps };
