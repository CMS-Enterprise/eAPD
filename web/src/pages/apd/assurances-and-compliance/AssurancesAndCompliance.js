import ChoiceList from '../../../components/ChoiceList';
import PropTypes from 'prop-types';
import React, { useEffect } from 'react';
import { useForm, Controller } from 'react-hook-form';
import { joiResolver } from '@hookform/resolvers/joi';
import { connect } from 'react-redux';
import {
  APD_TYPE,
  hitechAssurancesAndComplianceSchema,
  mmisAssurancesAndComplianceSchema
} from '@cms-eapd/common';

import { titleCase } from 'title-case';
import {
  setComplyingWithProcurement,
  setComplyingWithRecordsAccess,
  setComplyingWithSecurity,
  setComplyingWithSoftwareRights,
  setComplyingWithIndependentVV,
  setJustificationForProcurement,
  setJustificationForRecordsAccess,
  setJustificationForSecurity,
  setJustificationForSoftwareRights,
  setJustificationForIndependentVV
} from '../../../redux/actions/editApd';
import { Section, Subsection } from '../../../components/Section';
import TextArea from '../../../components/TextArea';
import regLinks from '../../../data/assurancesAndCompliance.yaml';
import { t } from '../../../i18n';
import {
  selectFederalCitations,
  selectAdminCheckEnabled,
  selectApdType
} from '../../../redux/selectors/apd.selectors';
import AlertMissingFFY from '../../../components/AlertMissingFFY';

const LinkOrText = ({ link, title }) => {
  if (!link) return title;

  return (
    <a href={link} target="_blank" rel="noopener noreferrer">
      {title}
    </a>
  );
};

LinkOrText.propTypes = {
  link: PropTypes.string,
  title: PropTypes.string.isRequired
};

LinkOrText.defaultProps = {
  link: null
};

const AssurancesAndCompliance = ({
  citations,
  complyingWithProcurement,
  complyingWithRecordsAccess,
  complyingWithSecurity,
  complyingWithSoftwareRights,
  complyingWithIndependentVV,
  justificationForProcurement,
  justificationForRecordsAccess,
  justificationForSecurity,
  justificationForSoftwareRights,
  justificationForIndependentVV,
  apdType,
  adminCheck
}) => {
  AssurancesAndCompliance.displayName = 'AssurancesAndCompliance';

  const {
    procurement = {},
    recordsAccess = {},
    softwareRights = {},
    security = {},
    independentVV = {}
  } = citations;
  const schema =
    apdType === APD_TYPE.HITECH
      ? hitechAssurancesAndComplianceSchema
      : mmisAssurancesAndComplianceSchema;

  const {
    control,
    trigger,
    clearErrors,
    setValue,
    formState: { errors }
  } = useForm({
    defaultValues: {
      procurement,
      recordsAccess,
      softwareRights,
      security,
      independentVV
    },
    mode: 'onChange',
    reValidateMode: 'onChange',
    resolver: joiResolver(schema)
  });

  useEffect(() => {
    if (adminCheck) {
      trigger();
    } else {
      clearErrors();
    }
  }, [adminCheck]); // eslint-disable-line react-hooks/exhaustive-deps

  function handleCheckChange(section, index, newValue) {
    switch (section) {
      case 'procurement':
        return complyingWithProcurement(index, newValue);
      case 'recordsAccess':
        return complyingWithRecordsAccess(index, newValue);
      case 'security':
        return complyingWithSecurity(index, newValue);
      case 'softwareRights':
        return complyingWithSoftwareRights(index, newValue);
      case 'independentVV':
        return complyingWithIndependentVV(index, newValue);
      default:
        return null;
    }
  }

  const handleExplanationChange = (section, index, value) => {
    switch (section) {
      case 'procurement':
        return justificationForProcurement(index, value);
      case 'recordsAccess':
        return justificationForRecordsAccess(index, value);
      case 'security':
        return justificationForSecurity(index, value);
      case 'softwareRights':
        return justificationForSoftwareRights(index, value);
      case 'independentVV':
        return justificationForIndependentVV(index, value);
      default:
        return null;
    }
  };

  return (
    <React.Fragment>
      <AlertMissingFFY />
      <Section id="assurances-compliance" resource="assurancesAndCompliance">
        <Subsection
          id="assurances-compliance-fed-citations"
          resource="assurancesAndCompliance.citations"
        >
          {Object.entries(regLinks[apdType]).map(([name, regulations]) => (
            <div
              key={name}
              className="ds-u-margin-top--6 ds-u-margin-bottom--0"
            >
              <h4 className="ds-h4">
                {titleCase(t(`assurancesAndCompliance.headings.${name}`))}
              </h4>
              {citations[name].map(({ title, checked, explanation }, index) => (
                <Controller
                  key={title}
                  name={`${name}.${index}.checked`}
                  control={control}
                  render={({ field: { onChange: radioOnChange } }) => (
                    <ChoiceList
                      label={
                        <legend className="ds-c-label">
                          Are you complying with{' '}
                          <strong>
                            <LinkOrText
                              link={regulations[title]}
                              title={title}
                            />
                          </strong>
                          ?
                        </legend>
                      }
                      name={`${name}.${index}.checked`}
                      value={checked}
                      choices={[
                        {
                          label: 'Yes',
                          value: 'yes',
                          checked: checked === true
                        },
                        {
                          label: 'No',
                          value: 'no',
                          checked: checked === false,
                          checkedChildren: (
                            <div className="ds-c-choice__checkedChild">
                              <Controller
                                name={`${name}.${index}.explanation`}
                                control={control}
                                render={({ field: { ...props } }) => (
                                  <TextArea
                                    {...props}
                                    label="Please explain"
                                    value={explanation}
                                    onChange={({ target: { value } }) => {
                                      setValue(
                                        `${name}.${index}.explanation`,
                                        value
                                      );
                                      handleExplanationChange(
                                        name,
                                        index,
                                        value
                                      );
                                      if (adminCheck) {
                                        trigger(`${name}.${index}.explanation`);
                                      }
                                    }}
                                    errorMessage={
                                      errors?.[name]?.[index]?.explanation
                                        ?.message
                                    }
                                    errorPlacement="bottom"
                                    rows={5}
                                  />
                                )}
                              />
                            </div>
                          )
                        }
                      ]}
                      type="radio"
                      size="small"
                      className="ds-u-margin-top--0"
                      onChange={e => {
                        const boolValue = e.target.value === 'yes';
                        radioOnChange(boolValue);
                        handleCheckChange(name, index, boolValue);
                        if (adminCheck) {
                          trigger(`${name}.${index}.checked`);
                          if (boolValue === false) {
                            trigger(`${name}.${index}.explanation`);
                          }
                        }
                      }}
                      errorMessage={errors?.[name]?.[index]?.checked?.message}
                      errorPlacement="bottom"
                    />
                  )}
                />
              ))}
            </div>
          ))}
        </Subsection>
      </Section>
    </React.Fragment>
  );
};

AssurancesAndCompliance.propTypes = {
  citations: PropTypes.object.isRequired,
  apdType: PropTypes.string.isRequired,
  adminCheck: PropTypes.bool.isRequired,
  complyingWithProcurement: PropTypes.func.isRequired,
  complyingWithRecordsAccess: PropTypes.func.isRequired,
  complyingWithSecurity: PropTypes.func.isRequired,
  complyingWithSoftwareRights: PropTypes.func.isRequired,
  complyingWithIndependentVV: PropTypes.func.isRequired,
  justificationForProcurement: PropTypes.func.isRequired,
  justificationForRecordsAccess: PropTypes.func.isRequired,
  justificationForSecurity: PropTypes.func.isRequired,
  justificationForSoftwareRights: PropTypes.func.isRequired,
  justificationForIndependentVV: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  citations: selectFederalCitations(state),
  apdType: (selectApdType(state) || '').toLowerCase(),
  adminCheck: selectAdminCheckEnabled(state)
});

const mapDispatchToProps = {
  complyingWithProcurement: setComplyingWithProcurement,
  complyingWithRecordsAccess: setComplyingWithRecordsAccess,
  complyingWithSecurity: setComplyingWithSecurity,
  complyingWithSoftwareRights: setComplyingWithSoftwareRights,
  complyingWithIndependentVV: setComplyingWithIndependentVV,
  justificationForProcurement: setJustificationForProcurement,
  justificationForRecordsAccess: setJustificationForRecordsAccess,
  justificationForSecurity: setJustificationForSecurity,
  justificationForSoftwareRights: setJustificationForSoftwareRights,
  justificationForIndependentVV: setJustificationForIndependentVV
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AssurancesAndCompliance);
export {
  AssurancesAndCompliance as plain,
  mapStateToProps,
  mapDispatchToProps,
  LinkOrText
};
