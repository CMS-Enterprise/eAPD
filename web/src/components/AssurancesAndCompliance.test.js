import { shallow } from 'enzyme';
import React from 'react';

import AssurancesAndCompliance from './AssurancesAndCompliance';

describe('AssurancesAndCompliance component', () => {
  test('renders correctly', () => {
    const component = shallow(<AssurancesAndCompliance />);
    expect(component).toMatchSnapshot();
  });
});
