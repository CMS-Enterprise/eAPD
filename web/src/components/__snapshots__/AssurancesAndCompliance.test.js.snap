// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssurancesAndCompliance component renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AssurancesAndCompliance />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Subsection
          open={false}
          resource="assurancesAndCompliance.procurement"
        >
          <div>
            ...
          </div>
        </Subsection>,
        <Subsection
          open={false}
          resource="assurancesAndCompliance.recordsAccess"
        >
          <div>
            ...
          </div>
        </Subsection>,
        <Subsection
          open={false}
          resource="assurancesAndCompliance.softwareRights"
        >
          <div>
            ...
          </div>
        </Subsection>,
        <Subsection
          open={false}
          resource="assurancesAndCompliance.security"
        >
          <div>
            ...
          </div>
        </Subsection>,
      ],
      "id": "assurances-compliance",
      "resource": "assurancesAndCompliance",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <div>
            ...
          </div>,
          "open": false,
          "resource": "assurancesAndCompliance.procurement",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "...",
          },
          "ref": null,
          "rendered": "...",
          "type": "div",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <div>
            ...
          </div>,
          "open": false,
          "resource": "assurancesAndCompliance.recordsAccess",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "...",
          },
          "ref": null,
          "rendered": "...",
          "type": "div",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <div>
            ...
          </div>,
          "open": false,
          "resource": "assurancesAndCompliance.softwareRights",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "...",
          },
          "ref": null,
          "rendered": "...",
          "type": "div",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <div>
            ...
          </div>,
          "open": false,
          "resource": "assurancesAndCompliance.security",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "...",
          },
          "ref": null,
          "rendered": "...",
          "type": "div",
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Subsection
            open={false}
            resource="assurancesAndCompliance.procurement"
          >
            <div>
              ...
            </div>
          </Subsection>,
          <Subsection
            open={false}
            resource="assurancesAndCompliance.recordsAccess"
          >
            <div>
              ...
            </div>
          </Subsection>,
          <Subsection
            open={false}
            resource="assurancesAndCompliance.softwareRights"
          >
            <div>
              ...
            </div>
          </Subsection>,
          <Subsection
            open={false}
            resource="assurancesAndCompliance.security"
          >
            <div>
              ...
            </div>
          </Subsection>,
        ],
        "id": "assurances-compliance",
        "resource": "assurancesAndCompliance",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <div>
              ...
            </div>,
            "open": false,
            "resource": "assurancesAndCompliance.procurement",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "...",
            },
            "ref": null,
            "rendered": "...",
            "type": "div",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <div>
              ...
            </div>,
            "open": false,
            "resource": "assurancesAndCompliance.recordsAccess",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "...",
            },
            "ref": null,
            "rendered": "...",
            "type": "div",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <div>
              ...
            </div>,
            "open": false,
            "resource": "assurancesAndCompliance.softwareRights",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "...",
            },
            "ref": null,
            "rendered": "...",
            "type": "div",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <div>
              ...
            </div>,
            "open": false,
            "resource": "assurancesAndCompliance.security",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "...",
            },
            "ref": null,
            "rendered": "...",
            "type": "div",
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
